### DISCUSSION TITLE
Support retrieving git metadata from export-subst file (`.git_archival.txt` pattern)

### DISCUSSION THREAD

**Discussion by nickstenning** [goreleaser/goreleaser#5549] [Ideas, Issue Triage, and General discussions]
### Is your feature request related to a problem? Please describe.

The problem this is trying to solve is that I'd like `goreleaser` to be the only way my project is built. That's tricky when packaging for Homebrew, for example, as we need to be able to build from a source tarball _without_ the `.git` directory. The problem is that `goreleaser build` doesn't allow me to provide the necessary git metadata any other way. (It seems like `GORELEASER_CURRENT_TAG` doesn't work for `build`?)

```console
$ GOTOOLCHAIN=go1.23.2 GORELEASER_CURRENT_TAG=v0.11.3 go run github.com/goreleaser/goreleaser/v2@latest build --skip=validate --clean --single-target --id cog --output cog
  • skipping validate...
  • cleaning distribution directory
  • loading environment variables
  • getting and validating git state
  ⨯ build failed after 0s                    error=current folder is not a git repository
exit status 1
```

### Describe the solution you'd like

What I'd like is for `goreleaser` to adopt the [same convention as Python's `setuptools_scm`](https://setuptools-scm.readthedocs.io/en/latest/usage/#git-archives), namely allowing Git users to specify an `export-subst` file which captures Git metadata in an archive tarball.

The idea is that if users create a `.gitattributes` file containing

```
.git_archival.txt  export-subst
```

and then a `.git_archival.txt` containing

```
node: $Format:%H$
node-date: $Format:%cI$
describe-name: $Format:%(describe:tags=true,match=*[0-9]*)$
```

then in the resulting archive tarball, the `.git_archival.txt` will contain a file that looks like this:

```
node: 72b293363490f32b754943f25ac82ee9976b0dc4
node-date: 2024-10-19T01:22:47+02:00
describe-name: v0.11.3
```

Python's `setuptools_scm` can [automatically detect this file](https://github.com/pypa/setuptools-scm/blob/be34dd11af6a5ef50f837d2c09dc98aaddd0a2dc/src/setuptools_scm/_file_finders/git.py#L103C5-L103C27) and use it to establish the relevant Git metadata for the build.

### Describe alternatives you've considered

Another option would be to allow `GORELEASER_CURRENT_TAG` and possibly other information like commit, commit date, etc., to be provided from environment variable overrides for `goreleaser build` invocations. This would work fine for us.

The advantage of the `.git_archival.txt` solution is that it's already supported by at least one other build tool, and it would be great to further establish it as a standard.

### Search

- [X] I did search for other open and closed issues before opening this

### Supporter

- [ ] I am a [sponsor](https://github.com/sponsors/caarlos0/) or a [Pro](https://goreleaser.com/pro) customer

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Additional context

Thank you for making goreleaser!

**Comment by caarlos0** [comment]
this sounds like an interesting idea...

although, quite a few things wouldn't work, of the top of my head, I can think of these:

- changelogs
- homebrew/scoop/etc URL detection 
- some environment variables, like tag summary, tag body, etc 

it seems that you are just calling `goreleaser build`, so I think what we can do is do not err if there's not `.git` and command is `build`.

That said, `GORELEASER_CURRENT_TAG` and `GORLEASER_PREVIOUS_TAG` are already allowed to be set. Not sure what other info you need, but I'm gonna play with this idea a bit before deciding which way to go.

that said, thanks for the issue, and sorry for the delay responding 🙏

**Comment by caarlos0** [comment]
I do not plan on implementing this any time soon, will convert to a discussion, we can keep talking about it there, and see what the community thinks as well.
